boot/bootutil/src/loader.c

```diff
@@ -1174,20 +1174,108 @@ boot_validated_swap_type(struct boot_loader_state *state,
+#define device_requires_erase(fa) (flash_area_erase_required(fa))
+#else
+#define device_requires_erase(fa) (true)
+#endif
+
 /**
- * Erases a region of flash.
+ * Erases a region of device that requires erase prior to write; does
+ * nothing on devices without erase.
  *
- * @param flash_area           The flash_area containing the region to erase.
+ * @param fap                   The flash_area containing the region to erase.
  * @param off                   The offset within the flash area to start the
- *                                  erase.
+ *                              erase.
  * @param sz                    The number of bytes to erase.
  *
  * @return                      0 on success; nonzero on failure.
  */
 int
 boot_erase_region(const struct flash_area *fap, uint32_t off, uint32_t sz)
 {
-    return flash_area_erase(fap, off, sz);
+    if (device_requires_erase(fap)) {
+        return flash_area_erase(fap, off, sz);
+    }
+    return 0;
+}
+
+/**
+ * Removes data from specified region either by writing 0x00 in place of
+ * data or by doing erase, if device has such hardware requirement.
+ *
+ * @param fa                    The flash_area containing the region to erase.
+ * @param off                   The offset within the flash area to start the
+ *                              erase.
+ * @param size                  The number of bytes to erase.
+ *
+ * @return                      0 on success; nonzero on failure.
+ */
+int
+boot_scramble_region(const struct flash_area *fa, uint32_t off, uint32_t size)
+{
+    int ret = 0;
+
+    if (device_requires_erase(fa)) {
+        return flash_area_erase(fa, off, size);
+    } else {
+        uint8_t buf[BOOT_MAX_ALIGN];
+        size_t size_done = 0;
+
+        memset(buf, flash_area_erased_val(fa), sizeof(buf));
+
+        while (size_done < size) {
+            ret = flash_area_write(fa, size_done + off, buf, sizeof(buf));
+            if (ret != 0) {
+                break;
+            }
+            size_done += sizeof(buf);
+        }
+    }
+    return ret;
+}
+
+/**
+ * Remove enough data from slot to mark is as unused
+ *
+ * @param fa        Pointer to flash area object for slot
+ *
+ * @return          0 on success; nonzero on failure.
+ */
+int
+boot_scramble_slot(const struct flash_area *fa)
+{
+    size_t size;
+
+    /* Whether device with erase or not, without minimal scramble
+     * removing deata in entire slot.
+     */
+#if !defined(MCUBOOT_MINIMAL_SCRAMBLE)
+    size = flash_area_get_size(fa);
+#else
+    if (device_requires_erase(fa)) {
+        int ret = 0;
+        struct flash_sector header;
+
+        ret = flash_area_get_sector(fa, 0, &header);
+        size = flash_sector_get_size(&header);
+
+        if (ret != 0) {
+            return ret;
+        }
+    } else {
+        size = MAX(sizeof(((struct image_header *)0)->ih_magic),
+                   BOOT_MAX_ALIGN);
+        size = (size + BOOT_MAX_ALIGN - 1) & ~(BOOT_MAX_ALIGN - 1);
```

Comment:
This should do the same for footer data because it might contain swap status or mcuboot flags for swapping images and those should always be cleared